<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Climatic-data-from-kaggle" data-toc-modified-id="Climatic-data-from-kaggle-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Climatic data from kaggle</a></span></li><li><span><a href="#Import-data" data-toc-modified-id="Import-data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Import data</a></span></li><li><span><a href="#Exploring-the-dataset" data-toc-modified-id="Exploring-the-dataset-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Exploring the dataset</a></span><ul class="toc-item"><li><span><a href="#Loking-for-NaN" data-toc-modified-id="Loking-for-NaN-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Loking for NaN</a></span><ul class="toc-item"><li><span><a href="#NaN-along-rows" data-toc-modified-id="NaN-along-rows-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>NaN along rows</a></span></li><li><span><a href="#NaN-along-columns" data-toc-modified-id="NaN-along-columns-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>NaN along columns</a></span></li></ul></li><li><span><a href="#Cleaning-the-data-set" data-toc-modified-id="Cleaning-the-data-set-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Cleaning the data set</a></span><ul class="toc-item"><li><span><a href="#Remove-the-repetitive-first-columns" data-toc-modified-id="Remove-the-repetitive-first-columns-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Remove the repetitive first columns</a></span></li><li><span><a href="#Get-Year-and-Months" data-toc-modified-id="Get-Year-and-Months-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Get Year and Months</a></span></li></ul></li><li><span><a href="#Some-statistics" data-toc-modified-id="Some-statistics-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Some statistics</a></span></li></ul></li></ul></div>

In [1]:
import sys
sys.path.append('../')

In [2]:
import pandas as pd
import src.functions_pipeline as fp
import src.import_kaggle_data as kd
from datetime import datetime

## Climatic data from kaggle  

To complement the information on the species I will use information on the air temperature of the last 40 years. This database is taken from www.kaggle.com.

[Daily_Temperature_Of_Major_Cities](https://www.kaggle.com/sudalairajkumar/daily-temperature-of-major-cities)


## Import data

In [ ]:
kd.download_dataset()

In [3]:
df = pd.read_csv("../Data/daily_temperature.csv",encoding = "ISO-8859-1", sep=';')

/home/anagarcia/anaconda3/envs/ironhack/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,24

In [4]:
df[:15]

Unnamed: 0           0              1                  2              3  \
0         city       Tokyo       New York        Mexico City         Mumbai   
1   city_ascii       Tokyo       New York        Mexico City         Mumbai   
2          lat      35.685        406.943            194.424         19.017   
3          lng   1.397.514       -739.249            -99.131         72.857   
4      country       Japan  United States             Mexico          India   
5         iso2          JP             US                 MX             IN   
6         iso3         JPN            USA                MEX            IND   
7   admin_name     Tōkyō       New York  Ciudad de México  Mahārāshtra   
8      capital     primary            NaN            primary          admin   
9   population  35676000.0     19354922.0         19028000.0     18978000.0   
10          id  1392685764     1840034016         1484247881     1356226629   
11    datetime           0              1                  2              3   
12  01/01/1980        4.16           1.48              11.35           23.4   
13  02/01/1980        4.06            1.0              10.01          22.67   
14  03/01/1980        7.12          -3.23               9.67           22.3   

             4           5           6            7              8  ...  \
0   São Paulo       Delhi    Shanghai      Kolkata    Los Angeles  ...   
1    Sao Paulo       Delhi    Shanghai      Kolkata    Los Angeles  ...   
2     -235.587       28.67     312.165       22.495        341.139  ...   
3      -46.625       77.23   1.214.365      883.247     -1.184.068  ...   
4       Brazil       India       China        India  United States  ...   
5           BR          IN          CN           IN             US  ...   
6          BRA         IND         CHN          IND            USA  ...   
7   São Paulo       Delhi    Shanghai  West Bengal     California  ...   
8        admin       admin       admin        admin            NaN  ...   
9   18845000.0  15926000.0  14987000.0   14787000.0     12815475.0  ...   
10  1076532519  1356872604  1156073548   1356060520     1840020491  ...   
11           4           5           6            7              8  ...   
12       17.86       13.74         9.1        20.47          14.24  ...   
13       18.65       12.85         6.4        19.37          13.77  ...   
14       18.42        10.5        5.43        18.71          12.08  ...   

              990         991         992           993           994  \
0       Bilāspur    Sargodha     Leipzig    Tinnevelly       Cancún   
1        Bilaspur    Sargodha     Leipzig    Tinnevelly        Cancun   
2         220.904     320.854     513.354        87.304         21.17   
3           82.16      72.675       12.41         77.69        -86.83   
4           India    Pakistan     Germany         India        Mexico   
5              IN          PK          DE            IN            MX   
6             IND         PAK         DEU           IND           MEX   
7   Chhattīsgarh      Punjab      Saxony  Tamil Nādu   Quintana Roo   
8             NaN       minor       minor           NaN         minor   
9        543454.0    542603.0    542529.0      542200.0      542043.0   
10     1356820491  1586696588  1276563678    1356037155    1484010310   
11            990         991         992           993           994   
12          22.14       11.54        0.67         24.05         21.21   
13          22.26       11.13       -0.88         24.19         23.29   
14          21.09       10.28        -4.3         24.39         24.04   

           995                     996              997                998  \
0     Yangzhou            Novokuznetsk          Latakia  Heroica Matamoros   
1     Yangzhou            Novokuznetsk          Latakia  Heroica Matamoros   
2         32.4                   53.75            35.54              25.88   
3       119.43                  87.115            35.78  

## Exploring the dataset

In [5]:
df.shape

(14896, 1001)

In [6]:
df.dtypes

Unnamed: 0    object
0             object
1             object
2             object
3             object
               ...  
995           object
996           object
997           object
998           object
999           object
Length: 1001, dtype: object

### Loking for NaN 

The first thing I want to do is to evaluate the number of NaN I have in the rows and columns to see if I should remove any of them because they have too many NaN.

#### NaN along rows 

In [7]:
null_rows = df.isnull().sum(axis = 1).apply (lambda y: y/df.shape[1]).sort_values(ascending = False)
null_rows[:5]

8       0.258741
7       0.017982
9       0.004995
5       0.000999
9929    0.000000
dtype: float64

The row 8 presents lowest percentage of NaN  with a 0.25%, so I will decide to leave it. 9929 (0.66 %) rows have all the data.

#### NaN along columns

In [8]:
null_col = df.isnull().sum().apply(lambda x: x/df.shape[0]).sort_values(ascending=False)
null_col[:5]

600    0.000134
32     0.000134
959    0.000134
601    0.000134
844    0.000067
dtype: float64

Overall, all columns are complete, so I won't delete any

### Cleaning the data set 

#### Remove the repetitive first columns

The first 11 rows have repeated information about the countries. I will leave only one, which I will also move to the top of the table. **I'll keep row 2 (with nidex 1) for the DataFrame header**

In [9]:
def get_header(df, iloc):
    '''
    Convert given row to column header for Pandas DataFrame
    
    Args:
        data(pd.DataFrame): working dataframe
        iloc(int): the iloc of the column we want to make heading
        
    Return:
        pd.DataFrame with the new heading determined
    '''
    headers = df.iloc[iloc] 
    df = pd.DataFrame(df.values[1:], columns=headers)
    df.columns = df.iloc[0, :]
    return df

In [10]:
df = get_header(df, 1)

In [11]:
df[:5]

0  city_ascii      Tokyo       New York  Mexico City  Mumbai  Sao Paulo  \
0  city_ascii      Tokyo       New York  Mexico City  Mumbai  Sao Paulo   
1         lat     35.685        406.943      194.424  19.017   -235.587   
2         lng  1.397.514       -739.249      -99.131  72.857    -46.625   
3     country      Japan  United States       Mexico   India     Brazil   
4        iso2         JP             US           MX      IN         BR   

0  Delhi   Shanghai  Kolkata    Los Angeles  ...  Bilaspur  Sargodha  Leipzig  \
0  Delhi   Shanghai  Kolkata    Los Angeles  ...  Bilaspur  Sargodha  Leipzig   
1  28.67    312.165   22.495        341.139  ...   220.904   320.854  513.354   
2  77.23  1.214.365  883.247     -1.184.068  ...     82.16    72.675    12.41   
3  India      China    India  United States  ...     India  Pakistan  Germany   
4     IN         CN       IN             US  ...        IN        PK       DE   

0  Tinnevelly  Cancun  Yangzhou  Novokuznetsk  Latakia  Heroica Matamoros  \
0  Tinnevelly  Cancun  Yangzhou  Novokuznetsk  Latakia  Heroica Matamoros   
1      87.304   21.17      32.4         53.75    35.54              25.88   
2       77.69  -86.83    119.43        87.115    35.78              -97.5   
3       India  Mexico     China        Russia    Syria             Mexico   
4          IN      MX        CN            RU       SY                 MX   

0  Goteborg  
0  Goteborg  
1     57.75  
2      12.0  
3    Sweden  
4        SE  

[5 rows x 1001 columns]

In [ ]:
df = fp.get_iloc(df)

In [ ]:
df

In [ ]:
df = fp.remove_rows(df)

In [ ]:
df[:5]

In [ ]:
df = fp.rename_columns(df, 'city_ascii', 'Date')
df

#### Get Year and Months 

In [ ]:
df = fp.get_month_year(df, 'Date')

In [ ]:
df[:5]

In [ ]:
df.to_csv("Temperature_Dataset.csv")

### Some statistics

In this pipeline I wl work with Spanish species, then te first step is select only this spanish cities. 

In [ ]:
df_spain = df[['Year', 'Month', 'Madrid', 'Barcelona', 'Malaga', 'Sevilla', 'Zaragoza', 'Valencia','Bilbao']]
df_spain.to_csv("Temperature_Spain2.csv")

In [ ]:
df_spain = pd.read_csv("Temperature_Spain2.csv")
df_spain[:5]

In [ ]:
df_spain = df_spain.drop(['Barcelona.1', 'Valencia', 'Unnamed: 0'], axis = 1)

In [ ]:
df_year = fp.mean_year(df_spain, 'Year')
df_month = fp.mean_month(df_spain, 'Month')

In [ ]:
df_year.to_csv("Temperature_Spain_year2.csv")
df_month.to_csv("Temperature_Spain_month2.csv")